bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java

```diff
@@ -53,38 +56,77 @@
     private static final String SELECTOR_FILTER_PROPERTY = "selectorFilter";
     private static final String INCLUDES_SUBCATEGORIES_PROPERTY = "includesSubCategories";
     private static final String ID_AND_URL_PATH_SEPARATOR = "|";
+    private static final String TEMPLATE_PREFIX = "{{";
+    private static final String TEMPLATE_SUFFIX = "}}";
 
     private String productUrlTemplate;
     private Pair<IdentifierLocation, ProductIdentifierType> productIdentifierConfig;
 
     private String categoryUrlTemplate;
-    private Pair<IdentifierLocation, CategoryIdentifierType> categoryIdentifierConfig;
+    private IdentifierLocation categoryIdentifierLocation;
+
+    private String queryParameterName;
 
     @Activate
     public void activate(UrlProviderConfiguration conf) {
         productUrlTemplate = conf.productUrlTemplate();
         productIdentifierConfig = Pair.of(conf.productIdentifierLocation(), conf.productIdentifierType());
 
         categoryUrlTemplate = conf.categoryUrlTemplate();
-        categoryIdentifierConfig = Pair.of(conf.categoryIdentifierLocation(), conf.categoryIdentifierType());
+        categoryIdentifierLocation = conf.categoryIdentifierLocation();
+
+        queryParameterName = conf.identifierQueryParamater();
     }
 
     @Override
     public String toProductUrl(SlingHttpServletRequest request, Page page, Map<String, String> params) {
         return toUrl(request, page, params, productUrlTemplate);
     }
 
+    @Override
+    public String toProductUrl(SlingHttpServletRequest request, Page page, String productIdentifier, MagentoGraphqlClient graphqlClient) {
+        ParamsBuilder params = new ParamsBuilder();
+        if (graphqlClient != null && StringUtils.isNotBlank(productIdentifier)) {
+            params.sku(productIdentifier);
+            ProductUrlParameterRetriever retriever = new ProductUrlParameterRetriever(graphqlClient);
+            retriever.setIdentifier(productIdentifier);
+            ProductInterface product = retriever.fetchProduct();
+            if (product != null) {
+                params.urlKey(product.getUrlKey());
+            } else {
+                LOGGER.debug("Could not generate product page URL for {}.", productIdentifier);
```

Comment:
Is a fall-through the right choice here? How would a URL to a product that cannot be fetched look like?

Reply:
Fall-through is only last option here and the returned URL will contain the placeholders. This behavior does not change we have it today. Without the url parameters we cannot do much here.